### PR TITLE
Efficient move operations for VFS files (SIRI-102)

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/storage/layer2/L3Uplink.java [hunk 1]
```diff
@@ -359,16 +359,32 @@ private boolean deleteHandler(VirtualFile file) {
         return false;
     }
 
-    private boolean moveHandler(VirtualFile file, VirtualFile newParent) {
+    private boolean canFastMoveHandler(VirtualFile file, VirtualFile newParent) {
```
Comment (tbiScireum) [inline src/main/java/sirius/biz/storage/layer2/L3Uplink.java]: javadoc ?: can move file into new parent directory if they belong both to the same space

File: src/main/java/sirius/biz/storage/layer3/uplink/util/UplinkConnectorPool.java [hunk 2]
```diff
@@ -8,9 +8,11 @@
 
 package sirius.biz.storage.layer3.uplink.util;
 
+import org.apache.commons.pool2.impl.DefaultPooledObject;
```
Comment (tbiScireum) [inline src/main/java/sirius/biz/storage/layer3/uplink/util/UplinkConnectorPool.java]: unncessary

File: src/main/java/sirius/biz/storage/layer3/uplink/util/Attempt.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    FIRST_ATTEMPT, RETRY;
+
+    /**
+     * Determines if the given exception should be thrown.
+     *
+     * @param exception the exception to check
+     * @return <tt>true</tt> if the exception should be thrown, <tt>false</tt> if it should be swallowed
+     */
+    public boolean shouldThrow(Exception exception) {
+        return this == RETRY || !((exception instanceof IOException) || (exception.getCause() instanceof IOException));
```
Comment (tbiScireum) [inline src/main/java/sirius/biz/storage/layer3/uplink/util/Attempt.java]: Wie wärs mit Exceptions.getRootCause ?

File: src/main/java/sirius/biz/storage/layer2/L3Uplink.java (hunk 1 above)
Comment (andyHa) [reply src/main/java/sirius/biz/storage/layer2/L3Uplink.java]: well, its a private method as simply provides the implementation as defined by MutableVirtualFile.withCanFastMoveHandler

File: src/main/java/sirius/biz/storage/layer3/uplink/ftp/FTPUplink.java [hunk 4]
```diff
@@ -185,25 +204,35 @@ private MutableVirtualFile wrap(VirtualFile parent, FTPFile file, String filenam
...
-        try (UplinkConnector<FTPClient> connector = connectorPool.obtain(ftpConfig)) {
-            FTPFile[] ftpFiles = list(connector.connector(),
-                                      file.parent().as(RemotePath.class),
-                                      ftpFile -> Strings.areEqual(ftpFile.getName(), file.name()));
-            if (ftpFiles.length == 1) {
-                file.attach(ftpFiles[0]);
-                return Optional.of(ftpFiles[0]);
-            } else {
-                return Optional.empty();
+        for (Attempt attempt : Attempt.values()) {
```
Comment (tbiScireum) [inline src/main/java/sirius/biz/storage/layer3/uplink/ftp/FTPUplink.java]: can we put that for loop & error handling in an extra method with a runnable or consumer as parameter? 
This looks like copy & paste on 3 different places. Maybe:

`attemptConnection(Consumer<UplinkConnector>)`

Comment (andyHa) [reply src/main/java/sirius/biz/storage/layer3/uplink/ftp/FTPUplink.java]: doesn't work within a lambda we need an inner return point

**Pull request merged**